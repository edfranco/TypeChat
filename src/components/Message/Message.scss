.messages-container {
    display: flex;
    flex-flow: column nowrap;
    overflow-y: auto;
    height: 80%;
    padding: 2em;
    .message:first-child {
        margin-top: auto !important;
        /* use !important to prevent breakage from child margin settings */
    }
    .message {
        margin: 1em 0;
        display: flex;
        width: 100%;

        .message-content {
            display: flex;
            flex-direction: row;
            justify-content: space-between;
        }

        .author-image-container {
            display: flex;
            justify-content: flex-end;
            margin-right: 1em;

            img {
                width: auto;
                height: 10vh;
                border-radius: 100%;
            }
        }

        .message-text {
            display: flex;
            flex-direction: column;
            justify-content: center;
            align-items: flex-start;
            max-width: 25rem;

            .message-info {
                display: flex;
                flex-direction: row;
                justify-content: flex-start;
                align-items: baseline;

                .message-author-name {
                    font-weight: bolder;
                    margin-right: 0.5em;
                }

                .message-time-stamp {
                    font-size: 0.6rem;
                }
            }

            .thread-info {
                display: flex;
                p {
                    margin: 1em 1em 0 0;
                    text-decoration: underline;
                    font-size: 0.5rem;
                    font-weight: bolder;
                }
                p:hover{
                    cursor: pointer;
                }
            }
        }
    }
}