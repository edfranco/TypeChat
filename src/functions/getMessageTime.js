const getMessageTime = () => {
    let date = new Date();
    let hour = date.getHours();
    let minutes = date.getMinutes();
    let isAfternoon = Boolean;

    if(minutes.toString().length < 2) {
        minutes = `${0}minutes`
    }

    if(hour > '12') {
        hour = `${hour - 12}`;
        isAfternoon = true;
    } else if (hour === '12') {
        hour = '12';
        isAfternoon = true;
    } else if (hour === '00') {
        hour = '12';
        isAfternoon = false;
    } else {
        hour = `${hour}`;
        isAfternoon = false
    }
    if(isAfternoon) {
        return `${hour}:${minutes}PM`;
    } else {
        return `${hour}:${minutes}AM`;
    }
}

export default getMessageTime;
