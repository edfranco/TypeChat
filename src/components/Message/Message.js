import { useState } from "react";
import Thread from "../Thread/Thread";
import getMessageTime from "../../functions/getMessageTime";
import "./Message.scss";

function Message(props) {
  const [thread, setThread] = useState([]);
  const [displayThread, setDisplayThread] = useState(false);
  const [displayThreadInput, setDisplayInput] = useState(false);
  const [threadMessageBody, setThreadMessageBody] = useState("");

  const handleSubmit = (event) => {
    event.preventDefault();
    setDisplayInput(!displayThreadInput);
    setDisplayThread(!displayThread);
    setThread([
      ...thread,
      {
        image: props.image,
        author: props.author,
        timeStamp: getMessageTime(),
        body: threadMessageBody,
      },
    ]);
  };

  const handleChange = (event) => {
    event.preventDefault();
    setThreadMessageBody(event.target.value);
  };

  return (
    <div className="message">
      <div className="message-content">
        <div className="author-image-container">
          <img src={props.image} />
        </div>
        <div className="message-text">
          <div className="message-info">
            <div className="message-author-name">
              <p>{props.author}</p>
            </div>
            <div className="message-time-stamp">
              <p>{props.timeStamp}</p>
            </div>
          </div>
          <div className="message-content">
            <p>{props.body}</p>
          </div>
          <div className="thread-info">
            <p
              className="reply-button"
              onClick={() => setDisplayInput(!displayThreadInput)}
            >
              Reply
            </p>
            {thread.length ? (
              <p onClick={() => setDisplayThread(!displayThread)}>
                View {thread.length} {thread.length > 1 ? "replies" : "reply"}
              </p>
            ) : null}
          </div>
          <form
            onSubmit={handleSubmit}
            style={
              displayThreadInput ? { display: "block" } : { display: "none" }
            }
          >
            <input value={threadMessageBody} onChange={handleChange} />
            <input type="submit" />
          </form>
        </div>
      </div>
      {displayThread ? (
        <Thread
          messages={thread}
          image={props.image}
          author={props.author}
          displayThread={displayThread}
          setDisplayThread={setDisplayThread}
          thread={thread}
          setThread={setThread}
        />
      ) : null}
    </div>
  );
}

export default Message;
