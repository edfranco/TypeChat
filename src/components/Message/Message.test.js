import { render, screen, cleanup } from "@testing-library/react";
import renderer from 'react-test-renderer';
import "@testing-library/jest-dom";

import Message from "./Message";
import getMessageTime from "../../functions/getMessageTime";

let testMessage = (
  <Message
    image=""
    author="Eduardo Franco"
    timeStamp={getMessageTime()}
    body="I am a message!"
  />
);

afterEach(() => {
  cleanup();
});

describe("Message component", () => {
  render(testMessage);

  const message = screen.getByTestId("message");
  const author = screen.getByTestId("message-author");
  const messageTime = screen.getByTestId("message-time");
  const messageBody = screen.getByTestId("message-body");
  const messageImage = screen.getByTestId("message-image");

  test("Message Rendering", () => {
    expect(message).toBeInTheDocument();
    expect(author).toBeInTheDocument();
    expect(messageTime).toBeInTheDocument();
    expect(messageBody).toBeInTheDocument();
    expect(messageImage).toBeInTheDocument();
  });

  test("Message author", () => {
    expect(author).toHaveTextContent("Eduardo Franco");
  });
  test("Message body", () => {
    expect(messageBody).toHaveTextContent("I am a message!");
  });
  test("Message timestamp", () => {
    expect(messageTime.textContent).toMatch(/(AM|PM)/i);
  });
});

it("Message renders correctly", () => {
  const tree = renderer
    .create(testMessage)
    .toJSON();
  expect(tree).toMatchSnapshot();
});
