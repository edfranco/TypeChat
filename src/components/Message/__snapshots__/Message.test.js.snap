// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Message renders correctly 1`] = `
<div
  className="message"
  data-testid="message"
>
  <div
    className="message-content"
  >
    <div
      className="author-image-container"
      data-testid="message-image"
    >
      <img
        src=""
      />
    </div>
    <div
      className="message-text"
    >
      <div
        className="message-info"
      >
        <div
          className="message-author-name"
          data-testid="message-author"
        >
          <p>
            Eduardo Franco
          </p>
        </div>
        <div
          className="message-time-stamp"
        >
          <p
            data-testid="message-time"
          >
            11:33PM
          </p>
        </div>
      </div>
      <div
        className="message-content"
        data-testid="message-body"
      >
        <p>
          I am a message!
        </p>
      </div>
      <div
        className="thread-info"
      >
        <p
          className="reply-button"
          onClick={[Function]}
        >
          Reply
        </p>
      </div>
      <form
        onSubmit={[Function]}
        style={
          Object {
            "display": "none",
          }
        }
      >
        <input
          onChange={[Function]}
          value=""
        />
        <input
          type="submit"
        />
      </form>
    </div>
  </div>
</div>
`;
